compiler/p/codegen/OMRTreeEvaluator.cpp

```diff
@@ -1101,9 +1101,88 @@ TR::Register *OMR::Power::TreeEvaluator::mloadiFromArrayEvaluator(TR::Node *node
     }
 }
 
+static TR::Register *mstoreiToArrayHelper(TR::Node *node, TR::CodeGenerator *cg,
+    TR::InstOpCode::Mnemonic reverseStoreOp, TR::InstOpCode::Mnemonic storeOp, TR::InstOpCode::Mnemonic splatOp,
+    int numElements)
+{
+    TR::Node *storeValNode = node->getSecondChild();
+
+    TR::Register *storeValReg = cg->evaluate(storeValNode);
+    TR::Register *tmpVRF = cg->allocateRegister(TR_VRF);
+    TR::Register *tmpGPR = NULL;
+
+    // set all but least significant bit of each array element to 0
+    if (numElements == 2) {
+        // since there is no splat immediate doubleword instruction, need to handle Long/DoubleVector mask separately
+        tmpGPR = cg->allocateRegister(TR_GPR);
+        generateTrg1ImmInstruction(cg, TR::InstOpCode::li, node, tmpGPR, 1);
+        generateTrg1Src1Instruction(cg, TR::InstOpCode::mtvsrd, node, tmpVRF, tmpGPR);
+        generateTrg1Src2ImmInstruction(cg, TR::InstOpCode::xxpermdi, node, tmpVRF, tmpVRF, tmpVRF, 0);
+    } else
+        generateTrg1ImmInstruction(cg, splatOp, node, tmpVRF, 1);
+
+    generateTrg1Src2Instruction(cg, TR::InstOpCode::xxland, node, tmpVRF, storeValReg, tmpVRF);
+
+    // pack doubleword-length elements into word-length elements (for Long/DoubleVector mask store)
+    if (numElements <= 2)
+        generateTrg1Src2Instruction(cg, TR::InstOpCode::vpksdus, node, tmpVRF, tmpVRF, tmpVRF);
+
+    // pack word-length elements into halfword-length elements (for Int/Float and Long/DoubleVector mask store)
+    if (numElements <= 4)
+        generateTrg1Src2Instruction(cg, TR::InstOpCode::vpkuwum, node, tmpVRF, tmpVRF, tmpVRF);
+
+    // pack halfworld-length elements into byte-length elements (for Short, Int/Float, and Long/DoubleVector mask store)
+    if (numElements <= 8)
+        generateTrg1Src2Instruction(cg, TR::InstOpCode::vpkuhum, node, tmpVRF, tmpVRF, tmpVRF);
+
+    auto ref = TR::LoadStoreHandlerImpl::generateMemoryReference(cg, node, numElements, true, 0);
+
+    if (cg->comp()->target().cpu.isAtLeast(OMR_PROCESSOR_PPC_P9) && numElements == 16)
+        generateMemSrc1Instruction(cg, TR::InstOpCode::stxvb16x, node, ref.getMemoryReference(), tmpVRF);
```

Comment:
I think would be better to do` if (numElements == 16)` and then assert `target().cpu.isAtLeast(OMR_PROCESSOR_PPC_P9)`

Reply:
There's already an assertion for that condition right before `mstoreiToArrayHelper()` is called for the `numElements == 16` case, so I just got rid of that condition from the if statement entirely.